### PR TITLE
arch: arm: be able to infer Z_ARCH_EXCEPT() for baseline SOCs

### LINKED ISSUE #18590: tests/kernel/fatal/kernel.common.stack_sentinel fails on FRDM-KW41Z
With commit: aed767a98c00bc11b5455fb60c3a3d8f0bda37c0

Running via:

`./scripts/sanitycheck -p frdm_kw41z --device-testing  --device-serial /dev/serial/by-id/usb-ARM_DAPLink_CMSIS-DAP_0201000028634e45000b5002673a0032a101000097969900-if01 -s tests/kernel/fatal/kernel.common.stack_sentinel`

Get the following output:

```
***** Booting Zephyr OS build v2.0.0-rc1-173-gaed767a98c00 *****
Running test suite fatal
===================================================================
starting test - test_fatal
test alt thread 1: generic CPU exception
FATAL: ***** HARD FAULT *****
FATAL: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x00000000
FATAL: r3/a4:  0x20000ac4 r12/ip:  0x20000158 r14/lr:  0x00000735
FATAL:  xpsr:  0x40000000
FATAL: Faulting instruction address (r15/pc): 0x20000ac4
FATAL: >>> ZEPHYR FATAL ERROR 0: CPU exception
FATAL: Current thread: 0x20000008 (unknown)
Caught system error -- reason 0
test alt thread 2: initiate kernel oops
FATAL: r0/a1:  0x00000003  r1/a2:  0x00000000  r2/a3:  0x00000003
FATAL: r3/a4:  0x00000000 r12/ip:  0x00000000 r14/lr:  0x00003991
FATAL:  xpsr:  0x01000000
FATAL: Faulting instruction address (r15/pc): 0x00000756
FATAL: >>> ZEPHYR FATAL ERROR 3: Kernel oops
FATAL: Current thread: 0x20000008 (unknown)
Caught system error -- reason 3
test alt thread 3: initiate kernel panic
FATAL: r0/a1:  0x00000004  r1/a2:  0x00000000  r2/a3:  0x00000004
FATAL: r3/a4:  0x00000000 r12/ip:  0x00000756 r14/lr:  0x00003991
FATAL:  xpsr:  0x01000000
FATAL: Faulting instruction address (r15/pc): 0x0000076e
FATAL: >>> ZEPHYR FATAL ERROR 4: Kernel panic
FATAL: Current thread: 0x20000008 (unknown)
Caught system error -- reason 4
test alt thread 4: fail assertion
ASSERTION FAIL [0] @ ZEPHYR_BASE/tests/kernel/fatal/src/main.c:123
	intentionally failed assertion
FATAL: r0/a1:  0x00000004  r1/a2:  0x0000007b  r2/a3:  0x000042d8
FATAL: r3/a4:  0x00000791 r12/ip:  0x0000076e r14/lr:  0x00000799
FATAL:  xpsr:  0x21000000
FATAL: Faulting instruction address (r15/pc): 0x00003c84
FATAL: >>> ZEPHYR FATAL ERROR 4: Kernel panic
FATAL: Current thread: 0x20000008 (unknown)
Caught system error -- reason 4
test stack sentinel overflow - timer irq
posting 1024 bytes of junk to stack...
FATAL: ***** HARD FAULT *****
FATAL: r0/a1:  0x00000002  r1/a2:  0xbbbbbbbb  r2/a3:  0xf0f0f0f0
FATAL: r3/a4:  0x200006d8 r12/ip:  0x20000000 r14/lr:  0x00001477
FATAL:  xpsr:  0x0100000f
FATAL: Faulting instruction address (r15/pc): 0x00002f30
FATAL: >>> ZEPHYR FATAL ERROR 0: CPU exception
FATAL: Current thread: 0x20000008 (unknown)
Caught system error -- reason 0
Wrong crash type got 0 expected 2
```

### PR DESCRIPTION
**Pull request by ioannisg**
This commit makes it possible to infer Z_ARCH_EXCEPT()
calls in SVCs that escalate to HardFault due to being
invoked from priority level equal or higher to the
interrupt priority level of the SVC Handler.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

Fixes #18590

### CODE REVIEW

**Comment by ioannisg** [comment]
This should be fixing #18590 
@galak could you, please, test?

**Review by galak — CHANGES_REQUESTED** [review]
Logic seems good, had a few comments on the code style.

File: arch/arm/core/cortex_m/fault.c [hunk 1]
```diff
@@ -609,6 +609,26 @@ static u32_t HardFault(z_arch_esf_t *esf, bool *recoverable)
...
+	 * which executes at priority equal to the SVC handler
+	 * priority. We handle the case of Kernel OOPS and Stack
+	 * Fail here.
+	 */
+		u32_t *ret_addr = (u32_t *)esf->basic.pc;
+		/* SVC is a 16-bit instruction. On a synchronous SVC
+		 * escalated to Hard Fault, the return address is the
+		 * next instruction, i.e. after the SVC.
+		 */
+		u16_t *svc_addr = (u16_t *)(((u8_t *)ret_addr) - 2);
```
[the remark this answers is not in the document]
Comment (ioannisg) [reply arch/arm/core/cortex_m/fault.c]: Sure, I can do that.

**Review by pizi-nordic — APPROVED** [review]
I would like to see @galak suggestion: https://github.com/zephyrproject-rtos/zephyr/pull/18713#discussion_r318606023 implemented here.

Everything else LGTM.

**Review by galak — APPROVED** [review]

**Review by ioannisg — COMMENTED** [review]
@galak updated code style based on your suggestion.
Could you please, confirm that this fixes the failing test on your platform?

It fixes the test on atsamr21_xpro.

File: arch/arm/core/cortex_m/fault.c [hunk 2]
```diff
@@ -609,6 +609,30 @@ static u32_t HardFault(z_arch_esf_t *esf, bool *recoverable)
...
+#define _SVC_OPCODE 0xDF00
+
+	u16_t fault_insn = *(ret_addr - 1);
+	if (((fault_insn & 0xff00) == _SVC_OPCODE) &&
+		((fault_insn & 0x00ff) == _SVC_CALL_RUNTIME_EXCEPT)) {
+
+		PR_EXC("Z_ARCH_EXCEPT with reason %x\n", esf->basic.r0);
+		reason = esf->basic.r0;
+	}
+#undef _SVC_OPCODE
```
[the remark this answers is not in the document]
Comment (ioannisg) [reply arch/arm/core/cortex_m/fault.c]: Not really - just thought it's ok to remove the definition, after it is not any more needed.

**Review by wentongwu — APPROVED** [review]

**Pull request merged**